instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ContainerResource.java

```diff
@@ -77,19 +77,83 @@ public static Resource get() {
   // Exists for testing
   static class Filesystem {
 
+    private static final Logger logger = Logger.getLogger(Filesystem.class.getName());
+
     boolean isReadable(Path path) {
       return Files.isReadable(path);
     }
 
     @MustBeClosed
     Stream<String> lines(Path path) throws IOException {
-      return Files.lines(path);
+
+      String osName = getOsName();
+      if (!(osName.equalsIgnoreCase("z/OS") || osName.equalsIgnoreCase("OS/390"))) {
+        return Files.lines(path);
+      } else {
+        return zosLines(path);
+      }
+    }
+
+    // This method reads the /proc system using the correct Charset for z/OS
+    @MustBeClosed
+    Stream<String> zosLines(Path path) throws IOException {
+
+      List<Charset> charsetsToTest = new ArrayList<Charset>();
+
+      // Since this class needs to be compatible with Java9&11, use the 'native.encoding' property
+      // to get the nativeCharset on java17+
+      // rather than System.console().charset().
+      if (System.getProperty("native.encoding") != null) {
+        Charset nativeCharset = Charset.forName(System.getProperty("native.encoding"));
+        if (nativeCharset != null) {
+          charsetsToTest.add(nativeCharset);
+        }
+      } else {
+        // On older javas that property won't exist but the default charset will be the native
+        // encoding.
+        charsetsToTest.add(Charset.defaultCharset());
+      }
+
+      // As a fallback value, add a hardcoded reference to IBM1047 (Canonical name: Cp1047).
+      Charset ibm1047 = Charset.forName("Cp1047");
+      if (ibm1047 != null) { // safety check
+        charsetsToTest.add(ibm1047);
+      }
+
+      // The odds of /proc being UTF-8 but the native encoding saying otherwise is extremely low
+      // but since UTF-8 is the standard it should be checked.
+      charsetsToTest.add(StandardCharsets.UTF_8);
+
+      IOException exception = null;
+      for (Charset charset : charsetsToTest) {
```

Comment:
could you elaborate under what circumstances you'd expect parsing with one charset to fail and proceed to the next one?

Reply:
Happy to. I'd expect this to fail on Z/OS if the JVM is configured with a non-default encoding on Z/OS. (I've had to do exactly that in our test environment)

In that case `Charset.defaultCharset()` will use whatever the user configured, if that is the wrong charset for reading `/proc` it will fall back to the next item which should pass since its hardcoded to Cp1047.

I did some tests with native.encoding on java17+ and couldn't change it from the default, but I'm not confident that I haven't missed something so its possible a similar scenario exists on java17+.

Reply:
Lets assume that the application runs on jdk11 or 8 that relies on `Charset.defaultCharset()` and uses `-Dfile.encoding=ISO-8859-1` (I think I saw somewhere a mention that websphere uses this). Now my guess would be that reading the file will succeed and the fallback to `Cp1047` won't happen. The content of the read file is not what it is supposed to be. If that is so then I think we need to come up with a different solution. Do you know is `com.ibm.jzos.ZUtil.getDefaultPlatformEncoding()` part of the java sdk on z/os? If it is we could add a compile dependency to https://central.sonatype.com/artifact/com.ibm.jzos/ibmjzos And use that to get the platform encoding.

Reply:
I tested `ISO-8859-1` and you're right, that is a fine catch that made it through all our automated tests. I also asked about `getDefaultPlatformEncoding()` but it turns out that also cannot guarantee correctness (especially if someone uses `setDefaultPlatformEncoding()` which is discouraged but has no enforcement against it).

The advice I was given was to keep it simple and just use `Cp1047`, as that would always be the case for `/proc` on ZOS. The other option was to read the [file tag](https://www.ibm.com/docs/en/semeru-runtime-ce-z/21.0.0?topic=reference-java-batch-toolkit-zos-jzos)  as `/proc` is all tagged with `Cp1047`. 

Which of these would you prefer?

Reply:
> The advice I was given was to keep it simple and just use Cp1047, as that would always be the case for /proc on ZOS.

That reads to me as you should always use `Cp1047` as this is known to be the encoding of that file and wouldn't need to look at `native.encoding` at all even when it is available. I'm not able to asses whether this is true or not. I'm willing to accept any solution that isn't obviously wrong. That includes unilaterally using `Cp1047`, falling back to `Cp1047` when `native.encoding` is not available or using `getDefaultPlatformEncoding()`.

Reply:
Ok, I'll unilaterally use `Cp1047`. I've heard from two different Z/OS developers that `/proc` is always encoded that way so it should be safe. And it will remove the need for a whole method of fallbacks.

Reply:
I ran the reworked version through our test system over the weekend it was fine so I've pushed it here, I've kept it in a separate commit for now for ease of reviewing.